modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/dht/topology/PartitionEvictionOrderTest.java

```diff
@@ -0,0 +1,199 @@
+        super.afterTest();
+
+        stopAllGrids();
+
+        cleanPersistenceDir();
+    }
+
+    /**
+     * Tests that {@link CacheRebalanceMode#SYNC} caches are evicted at first.
+     */
+    @Test
+    @WithSystemProperty(key = IgniteSystemProperties.IGNITE_EVICTION_PERMITS, value = "1")
+    public void testSyncCachesEvictedAtFirst() throws Exception {
+        IgniteEx node0 = startGrid(0);
+
+        node0.cluster().active(true);
+
+        IgniteEx node1 = startGrid(1);
+
+        node0.cluster().setBaselineTopology(node1.cluster().topologyVersion());
+
+        GridCacheAdapter<Object, Object> utilCache0 = grid(0).context().cache().internalCache(CU.UTILITY_CACHE_NAME);
+
+        IgniteCache<Object, Object> cache = node0.getOrCreateCache(DEFAULT_CACHE_NAME);
+
+        for (int i = 0; i < 100_000; i++) {
+            utilCache0.put(i, i);
+
+            cache.put(i, i);
+        }
+
+        awaitPartitionMapExchange();
+
+        stopGrid(0);
+
+        GridCacheAdapter<Object, Object> utilCache1 = grid(1).context().cache().internalCache(CU.UTILITY_CACHE_NAME);
+
+        IgniteCache<Object, Object> cache2 = grid(1).cache(DEFAULT_CACHE_NAME);
+
+        for (int i = 0; i < 200_000; i++) {
+            try {
+                cache2.put(i, i + 1);
+
+                utilCache1.put(i, i + 1);
+            }
+            catch (IgniteCheckedException e) {
+                e.printStackTrace();
+            }
+        }
+
+        CountDownLatch latch = new CountDownLatch(1);
+
+        TestDependencyResolver resolver = new TestDependencyResolver(new DependencyResolver() {
+            @Override public <T> T resolve(T instance) {
+                if (instance instanceof PartitionsEvictManager.BucketQueue)
+                    return (T)(new TestBucketQueue(1, latch));
+
+                return instance;
+            }
+        });
+
+        startGrid(0, resolver);
+
+        //We start not from 0 because the first partition that comes to queue is from async cache.
+        for (int i = 1; i < 101; i++)
+            assertEquals(CU.UTILITY_CACHE_GROUP_ID, evictionOrder.get(i).intValue());
+
+    }
+
+    /**
+     * Test Bucket queue that has the same logic as {@link PartitionsEvictManager.BucketQueue},
+     * but with a latch in {@code poll} method to wait for queue to be filled with partitions.
+     */
+    private class TestBucketQueue extends PartitionsEvictManager.BucketQueue {
+        /**
+         * Latch is needed for waiting queue to be fillled with partitions.
+         */
+        private CountDownLatch latch;
+
+        /**
+         * @param buckets Number of buckets.
+         */
+        TestBucketQueue(int buckets, CountDownLatch latch) {
+            super(buckets);
+
+            this.latch = latch;
+        }
+
+        /** {@inheritDoc} */
+        @Override PartitionsEvictManager.PartitionEvictionTask poll(int bucket) {
+            try {
+                if (!latch.await(5, TimeUnit.SECONDS))
```

Comment:
Looks a bit fragile. 
Can we wait until all partitions from sys-cache and default cache are scheduled for eviction instead? We can check that size of the queue is equal to total number of partitions.